kv.go

```diff
@@ -0,0 +1,387 @@
+	}
+	return kv, nil
+}
+
+type kvs struct {
+	name   string
+	stream string
+	pre    string
+	nc     *Conn
+	js     *js
+}
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(key string) (value []byte, revision uint64, err error) {
+	// Build by hand since simple and avoids stdlib JSON.
+	var b strings.Builder
+	b.WriteString("{\"last_by_subj\":\"")
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+	b.WriteString("\"}")
+
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return nil, 0, err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	// Send request.
+	subj := kv.js.apiSubj(fmt.Sprintf(apiMsgGetT, kv.stream))
+	r, err := kv.nc.RequestWithContext(o.ctx, subj, []byte(b.String()))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// FIXME(dlc) - Be good to avoid stdlib JSON when possible.
+	// Maybe: https://github.com/goccy/go-json
+	var resp apiMsgGetResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return nil, 0, err
+	}
+	if resp.Error != nil {
+		return nil, 0, errors.New(resp.Error.Description)
+	}
+
+	return resp.Message.Data, resp.Message.Sequence, nil
+}
+
+// Put will place the new value for the key into the store.
+func (kv *kvs) Put(key string, value []byte) (revision uint64, err error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	pa, err := kv.js.Publish(b.String(), value)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Create will add the key/value pair iff it does not exist.
+func (kv *kvs) Create(key string, value []byte) (revision uint64, err error) {
+	return kv.Update(key, value, 0)
+}
+
+// Update will update the value iff the latest revision matches.
+func (kv *kvs) Update(key string, value []byte, revision uint64) (uint64, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m := Msg{Subject: b.String(), Header: Header{}, Data: value}
+	m.Header.Set(ExpectedLastSubjSeqHdr, strconv.FormatUint(revision, 10))
+
+	pa, err := kv.js.PublishMsg(&m)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Delete the key and all revisions.
+func (kv *kvs) Delete(key string) error {
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	// Build by hand since simple and avoids stdlib JSON.
+	var b strings.Builder
+	b.WriteString("{\"filter\":\"")
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+	b.WriteString("\"}")
+
+	// Send request.
+	subj := kv.js.apiSubj(fmt.Sprintf(apiStreamPurgeT, kv.stream))
+	r, err := kv.nc.RequestWithContext(o.ctx, subj, []byte(b.String()))
+	if err != nil {
+		return err
+	}
+	var resp streamPurgeResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return err
+	}
+	if resp.Error != nil {
+		return errors.New(resp.Error.Description)
+	}
+	return nil
+}
+
+// List will return all values for the key.
+func (kv *kvs) List(key string) ([]*KeyValueEntry, error) {
```

Comment:
I'd have thought you'd make this an iterator or channel since you made the same for things like StreamNames? I find that completely unusable - but it seems it would be more consistent with the design of this library?

Reply:
Another good discussion, these seem more like "one shot" vs paging and I was taking a lead from how Fetch() was done which returns []*nats.Msg.

Reply:
yeah, right now I read the entire stream and put all subjects (keys) into a set, so at the end I have a set of the keys. Clearly not efficient.

Reply:
I don't see people doing crazy big histories do we?

Reply:
I considered capping it to some number below 100

Reply:
I am ok with that.. 64 to start?

Reply:
Yes